lib/elixir/lib/enum.ex

```diff
@@ -717,30 +717,61 @@ defmodule Enum do
   """
   @spec fetch(t, integer) :: {:ok, element} | :error
   def fetch(enumerable, index)
+
+  def fetch(enumerable, index) when is_integer(index) and index < 0 do
+    case Enumerable.count(enumerable) do
+      {:ok, count} when (count + index) < 0 ->
+        :error
+
+      {:ok, count} ->
+        fetch(enumerable, count + index)
+
+      {:error, _} ->
+        enumerable
+        |> reverse([])
+        |> fetch_list(abs(index + 1))
+    end
+  end
+
+  def fetch(enumerable, index)
       when is_list(enumerable) and is_integer(index) and index >= 0 do
-    do_fetch(enumerable, index)
+    fetch_list(enumerable, index)
   end
 
   def fetch(enumerable, index) when is_integer(index) and index >= 0 do
-    res =
-      Enumerable.reduce(enumerable, {:cont, 0}, fn(entry, acc) ->
-        if acc == index do
-          {:halt, entry}
-        else
-          {:cont, acc + 1}
-        end
-      end)
+    # this avoids unnecessarily traversing an enumerable that implements Enumerable.count/1
+    # given an invalid index
+    count_res =
+      case Enumerable.count(enumerable) do
+        {:ok, count} when (count - 1 - index) < 0 ->
+          :error
+        {:ok, _} ->
+          {:ok, Enumerable}
```

Comment:
@josevalim I'm refactoring this code, because there's still room to speed things up. I would like to know at this point, where the answer is {:ok, _}, what is the name of the target module of this protocol. so I avoid doing Enumerable.reduce, and just call module.reduce